From dfac1ba628ec6e77ff8af0526994f7082848162a Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:53:58 -0400
Subject: [fs] ext4: fix del_timer() misuse for ->s_err_report

Message-id: <1414058125-4183-85-git-send-email-lczerner@redhat.com>
Patchwork-id: 98158
O-Subject: [RHEL 7.1 PATCH v2 084/171] ext4: fix del_timer() misuse for ->s_err_report
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Al Viro <viro@ZenIV.linux.org.uk>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit 9105bb149bbbc555d2e11ba5166dfe7a24eae09e

 That thing should be del_timer_sync(); consider what happens
 if ext4_put_super() call of del_timer() happens to come just as it's
 getting run on another CPU.  Since that timer reschedules itself
 to run next day, you are pretty much guaranteed that you'll end up
 with kfree'd scheduled timer, with usual fun consequences.  AFAICS,
 that's -stable fodder all way back to 2010... [the second del_timer_sync()
 is almost certainly not needed, but it doesn't hurt either]

 Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Cc: stable@vger.kernel.org
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/super.c b/fs/ext4/super.c
index 287bc34..b8a5088 100644
--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -794,7 +794,7 @@ static void ext4_put_super(struct super_block *sb)
  }
 
  ext4_es_unregister_shrinker(sbi);
- del_timer(&sbi->s_err_report);
+ del_timer_sync(&sbi->s_err_report);
  ext4_release_system_zone(sb);
  ext4_mb_release(sb);
  ext4_ext_release(sb);
@@ -4198,7 +4198,7 @@ failed_mount_wq:
  }
 failed_mount3:
  ext4_es_unregister_shrinker(sbi);
- del_timer(&sbi->s_err_report);
+ del_timer_sync(&sbi->s_err_report);
  if (sbi->s_flex_groups)
   ext4_kvfree(sbi->s_flex_groups);
  percpu_counter_destroy(&sbi->s_freeclusters_counter);
-- 
1.7.1